### PR TITLE
CNDE-1849: Refactor Investigation Service to use Kafka updates for Page Case Answers

### PR DESCRIPTION
**Pull request by sveselev**
## Notes

This PR improves overall performance by reducing the processing time in the Investigation Service. Direct interactions with RDB database replaces by delete/update via Kafka topic for `nrt_page_case_answer`.

## JIRA

- **Related story**: [CNDE-1849](https://cdc-nbs.atlassian.net/browse/CNDE-1849)

## Checklist

- [x] PR focuses on a single story.
- [x] New unit tests added and ensured they pass.
- [x] Service has been tested in local and it works as expected.
- [ ] Documentation has been updated for this code change (if needed).
- [x] Code follows the Java Coding Conventions (https://www.oracle.com/java/technologies/javase/codeconventions-programmingpractices.html).

## Types of changes

What types of changes does this PR introduces?

- [ ] Bugfix
- [ ] New feature
- [ ] Breaking change

## Testing

- [x] Does this PR has >90% code coverage?
- [ ] Is the screenshot attached for code coverage?
- [x] Does the `gradle build` pass in your local? 
- [ ] Is the `gradle build` logs attached?

### CODE REVIEW

File: common-util/src/main/java/gov/cdc/etldatapipeline/commonutil/json/CustomJsonGeneratorImpl.java [hunk 1]
```diff
@@ -41,9 +41,9 @@ private static ArrayNode generateFieldsArray(Object model) {
 
                 fieldNode.put("type", getType(field.getType().getSimpleName().toLowerCase()));
                 fieldNode.put("optional", (!fieldName.equals("public_health_case_uid")
-                        && !fieldName.equals("observation_uid") && !fieldName.equals("organization_uid")
-                        && !fieldName.equals("patient_uid") && !fieldName.equals("provider_uid")
-                        && !fieldName.equals("notification_uid")));
+                        && !fieldName.equals("act_uid") && !fieldName.equals("observation_uid")
```
Comment (alog-enquizit) [inline common-util/src/main/java/gov/cdc/etldatapipeline/commonutil/json/CustomJsonGeneratorImpl.java:44]: Suggestion: We can put all the primary keys in a list and use a not contains method.

File: investigation-service/src/main/java/gov/cdc/etldatapipeline/investigation/util/ProcessInvestigationDataUtil.java [hunk 2]
```diff
@@ -225,11 +224,15 @@ private void transformInvestigationConfirmationMethod(String investigationConfir
             investigationConfirmation.setConfirmationMethodTime(
                     confirmationMethodTime == null ? phcLastChgTime : confirmationMethodTime);
 
+            // Tombstone message to delete all confirmation methods for specified phc uid
+            String jsonKey = jsonGenerator.generateStringJson(new InvestigationConfirmationMethodUidKey(publicHealthCaseUid));
+            kafkaTemplate.send(investigationConfirmationOutputTopicName, jsonKey, null);
```
Comment (alog-enquizit) [inline investigation-service/src/main/java/gov/cdc/etldatapipeline/investigation/util/ProcessInvestigationDataUtil.java:229]: Good 👍

**Review by alog-enquizit — APPROVED** [review]
lgtm 👍

File: investigation-service/src/main/java/gov/cdc/etldatapipeline/investigation/util/ProcessInvestigationDataUtil.java [hunk 3]
```diff
@@ -245,29 +248,40 @@ private void processInvestigationPageCaseAnswer(String investigationCaseAnswer,
             JsonNode investigationCaseAnswerJsonArray = parseJsonArray(investigationCaseAnswer);
 
             Long actUid = investigationCaseAnswerJsonArray.get(0).get("act_uid").asLong();
-            List<InvestigationCaseAnswer> investigationCaseAnswerList = new ArrayList<>();
+            List<PageCaseAnswer> pageCaseAnswerList = new ArrayList<>();
+
+            // Tombstone message to delete all page case answers for specified actUid
+            PageCaseAnswerUidKey pageCaseAnswerUidKey = new PageCaseAnswerUidKey(actUid);
+            String jsonKey = jsonGenerator.generateStringJson(pageCaseAnswerUidKey);
+            kafkaTemplate.send(pageCaseAnswerOutputTopicName, jsonKey, null);
```
Comment (upasanapattnaik-eq) [inline investigation-service/src/main/java/gov/cdc/etldatapipeline/investigation/util/ProcessInvestigationDataUtil.java:256]: 👍

File: liquibase-service/src/main/resources/db/rdb_modern/routines/005-sp_nrt_investigation_postprocessing-003.sql [hunk 4]
```diff
@@ -0,0 +1,637 @@
...
+        SET @proc_step_name = 'UPDATE CONFIRMATION_METHOD_GROUP';
+        SET @proc_step_no = 5;
+
+        delete dbo.CONFIRMATION_METHOD_GROUP
+        where investigation_key in
+              (select investigation_key from dbo.INVESTIGATION where case_uid in
+                    (select value FROM STRING_SPLIT(@id_list, ','))
+              )
+
+        insert into dbo.CONFIRMATION_METHOD_GROUP ([INVESTIGATION_KEY],[CONFIRMATION_METHOD_KEY],[CONFIRMATION_DT])
```
Comment (upasanapattnaik-eq) [inline liquibase-service/src/main/resources/db/rdb_modern/routines/005-sp_nrt_investigation_postprocessing-003.sql:533]: After this update, will we still require the max date for case answer code?

**Review by upasanapattnaik-eq — APPROVED** [review]
LGTM!

File: liquibase-service/src/main/resources/db/rdb_modern/routines/005-sp_nrt_investigation_postprocessing-003.sql (hunk 4 above)
Comment (sveselev) [reply liquibase-service/src/main/resources/db/rdb_modern/routines/005-sp_nrt_investigation_postprocessing-003.sql:533]: > After this update, will we still require the max date for case answer code?

Actually, we will not. Would you suggest to rollback those changes too?

Comment (upasanapattnaik-eq) [reply liquibase-service/src/main/resources/db/rdb_modern/routines/005-sp_nrt_investigation_postprocessing-003.sql:533]: It adds a potential processing overhead if it is a subquery. We can take up the evaluation and rollback as part of another story.

**Pull request merged**